### PR TITLE
NODE-2627 State Hash calculation changes

### PR DESCRIPTION
**Pull request by phearnot**

### CODE REVIEW

File: grpc-server/src/main/scala/com/wavesplatform/events/events.scala [hunk 1]
```diff
@@ -430,27 +422,37 @@ object StateUpdate {
...
+    }
+
+    val cancelledLeaseUpdates = snapshot.cancelledLeases.map { case (id, _) =>
+      val si = snapshot.newLeases.get(id).orElse(blockchain.leaseDetails(id).map(_.static))
+      LeaseUpdate(
+        id,
+        LeaseStatus.Inactive,
+        si.fold(0L)(_.amount.value),
+        si.fold(PublicKey(new Array[Byte](32)))(_.sender),
+        si.fold(PublicKey(new Array[Byte](32)).toAddress)(_.recipientAddress),
```
Comment (vsuharnikov) [inline grpc-server/src/main/scala/com/wavesplatform/events/events.scala:444]:
```suggestion
        si.fold(PublicKeys.EmptyPublicKey)(_.sender),
        si.fold(PublicKeys.EmptyPublicKey.toAddress)(_.recipientAddress),
```

File: node/src/main/scala/com/wavesplatform/api/common/CommonAccountsApi.scala [hunk 2]
```diff
@@ -188,8 +188,8 @@ object CommonAccountsApi {
         leaseId,
         ld.sourceId,
         ld.sender.toAddress,
-        blockchain.resolveAlias(ld.recipient).orElse(resolveDisabledAlias(leaseId)).explicitGet(),
-        ld.amount,
+        blockchain.resolveAlias(ld.recipientAddress).orElse(resolveDisabledAlias(leaseId)).explicitGet(),
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/api/common/CommonAccountsApi.scala]:
```suggestion
        ld.recipientAddress,
```
`ld.recipientAddress` is `Address`, so `blockchain.resolveAlias(ld.recipientAddress)` returns `Right` and we won't fall into `resolveDisabledAlias(leaseId)`

**Review by xrtm000 — APPROVED** [review]

**Review by ivan-mashonskiy — APPROVED** [review]

File: node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala [hunk 3]
```diff
@@ -484,15 +483,16 @@ class RocksDBWriter(
...
-      val allLeases    = txLeases ++ snapshot.leaseStates.collect { case (id, lease) if !txLeaseIdSet.contains(id) => (None, id, lease) }
-      allLeases.foreach { case (txOpt, id, lease) =>
-        val details = lease.toDetails(this, txOpt, leaseDetails(id))
-        rw.put(Keys.leaseDetails(id)(height), Some(Right(details)))
+      for ((id, li) <- snapshot.newLeases) {
+        rw.put(Keys.leaseDetails(id)(height), Some(LeaseDetails(li, snapshot.cancelledLeases.getOrElse(id, LeaseDetails.Status.Active))))
+        expiredKeys ++= updateHistory(rw, Keys.leaseDetailsHistory(id), threshold, Keys.leaseDetails(id))
+      }
+
+      for ((id, status) <- snapshot.cancelledLeases if !snapshot.newLeases.contains(id)) {
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala:491]: Should we add the `!snapshot.newLeases.contains(id)` filter here? What if there are LeaseTransaction and LeaseCancelTransaction for same leasing in one block?

File: grpc-server/src/main/scala/com/wavesplatform/events/events.scala (hunk 1 above)
Comment (phearnot) [reply grpc-server/src/main/scala/com/wavesplatform/events/events.scala:444]: `grpc-server` module does not depend on `ride-runner`, so `PublicKeys` is inaccessible from here.

File: node/src/main/scala/com/wavesplatform/state/BlockchainUpdaterImpl.scala [hunk 4]
```diff
@@ -431,24 +429,21 @@ class BlockchainUpdaterImpl(
...
+      recipient <- rocksdb.resolveAlias(lt.recipientAddress).toSeq
+      portfolios = Map(
+        lt.sender.toAddress -> Portfolio(0, LeaseBalance(0, -lt.amount.value)),
+        recipient           -> Portfolio(0, LeaseBalance(-lt.amount.value, 0))
+      )
+      leaseStates = Map(
+        id -> LeaseDetails.Status.Expired(height)
+      )
+      snapshot = StateSnapshot.build(rocksdb, portfolios, cancelledLeases = leaseStates).explicitGet()
+    } yield id -> snapshot
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/state/BlockchainUpdaterImpl.scala]:
```suggestion
    for ((id, lt) <- leaseTransactions) yield {
      val portfolios = Map(
        lt.sender.toAddress -> Portfolio(0, LeaseBalance(0, -lt.amount.value)),
        lt.recipientAddress -> Portfolio(0, LeaseBalance(-lt.amount.value, 0))
      )
      val leaseStates = Map(
        id -> LeaseDetails.Status.Expired(height)
      )
      val snapshot = StateSnapshot.build(rocksdb, portfolios, cancelledLeases = leaseStates).explicitGet()
      id -> snapshot
    }
```

File: node/src/main/scala/com/wavesplatform/state/diffs/DiffsCommon.scala [hunk 5]
```diff
@@ -186,7 +185,7 @@ object DiffsCommon {
     val allowedTs = blockchain.settings.functionalitySettings.allowMultipleLeaseCancelTransactionUntilTimestamp
     for {
       lease     <- blockchain.leaseDetails(leaseId).toRight(GenericError(s"Lease with id=$leaseId not found"))
-      recipient <- blockchain.resolveAlias(lease.recipient)
+      recipient <- blockchain.resolveAlias(lease.recipientAddress)
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/state/diffs/DiffsCommon.scala]:
```suggestion
```
Can't suggest a change for modified and unmodified lines in GitHub :( See below.

File: node/src/main/scala/com/wavesplatform/state/diffs/DiffsCommon.scala [hunk 6]
```diff
@@ -200,14 +199,13 @@ object DiffsCommon {
             s"time=$time > allowMultipleLeaseCancelTransactionUntilTimestamp=$allowedTs"
         )
       )
-      senderPortfolio    = Map[Address, Portfolio](sender.toAddress -> Portfolio(-fee, LeaseBalance(0, -lease.amount)))
-      recipientPortfolio = Map(recipient -> Portfolio(0, LeaseBalance(-lease.amount, 0)))
-      actionInfo         = lease.copy(status = LeaseDetails.Status.Cancelled(blockchain.height, Some(cancelTxId)))
+      senderPortfolio    = Map[Address, Portfolio](sender.toAddress -> Portfolio(-fee, LeaseBalance(0, -lease.amount.value)))
+      recipientPortfolio = Map(recipient -> Portfolio(0, LeaseBalance(-lease.amount.value, 0)))
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/state/diffs/DiffsCommon.scala]:
```suggestion
      recipientPortfolio = Map(lease.recipientAddress -> Portfolio(0, LeaseBalance(-lease.amount.value, 0)))
```

File: node/src/main/scala/com/wavesplatform/state/patch/CancelAllLeases.scala [hunk 7]
```diff
@@ -1,23 +1,19 @@
...
 
   private[patch] case class CancelledLeases(balances: Map[Address, LeaseBalance], cancelledLeases: Seq[LeaseData]) {
     private[this] val height: Int = patchHeight.getOrElse(0)
-    val leaseStates: Map[ByteStr, LeaseSnapshot] = cancelledLeases.map { data =>
-      val sender    = PublicKey(ByteStr.decodeBase58(data.senderPublicKey).get)
-      val recipient = Address.fromString(data.recipient).explicitGet()
-      val id        = ByteStr.decodeBase58(data.id).get
-      (id, LeaseSnapshot(sender, recipient, data.amount, status = LeaseDetails.Status.Expired(height)))
+    val leaseStates: Map[ByteStr, LeaseDetails.Status.Inactive] = cancelledLeases.map { data =>
+      (ByteStr.decodeBase58(data.id).get, LeaseDetails.Status.Cancelled(height, None))
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/state/patch/CancelAllLeases.scala]: The previous status is `Expired`. Now it is `Cancelled`. Is this intentional?

**Review by vsuharnikov — APPROVED** [review]

**Review by DrBlast — APPROVED** [review]

**Pull request merged**